maro/simulator/scenarios/supply_chain/units/storage.py

```diff
@@ -1,51 +1,219 @@
+            average_space = remaining_space // sku_num
+            for sku_id, upper_bound in upper_bound_dict.items():
+                if upper_bound is None:
+                    sku_num -= 1
+                    if sku_num > 0:
+                        upper_bound_dict[sku_id] = average_space
+                        remaining_space -= average_space
+                    else:
+                        # In the case the initial remaining space is not divisible.
+                        upper_bound_dict[sku_id] = remaining_space
+
+        self.data_model.initialize(
+            capacity=self.capacity,
+            remaining_space=self.remaining_space,
+            product_list=[sku_id for sku_id in self._product_level.keys()],
+            product_quantity=[n for n in self._product_level.values()],
+        )
+
+    def get_product_quantity(self, product_id: int) -> int:
+        """Get product quantity in storage.
 
         Args:
-            product_quantities (Dict[int, int]): Dictionary of product id and quantity need to add to storage.
-            all_or_nothing (bool): Failed if all product cannot be added, or add as many as it can. Default is True.
+            product_id (int): Product to check.
 
         Returns:
-            Dict[int, int]: Dictionary of product id and quantity success added.
+            int: Available quantity of product.
         """
-        if all_or_nothing and self.remaining_space < sum(product_quantities.values()):
-            return {}
+        return self._product_level[product_id]
 
-        unloaded_quantities: Dict[int, int] = {}
+    def get_product_max_remaining_space(self, product_id: int) -> int:
+        # ! Currently the upper bound is only used in Manufacture Unit.
+        remaining_space = self._remaining_space_dict[self._product2storage[product_id]]
 
-        for product_id, quantity in product_quantities.items():
-            unload_quantity = min(self.remaining_space, quantity)
+        upper_bound = self._storage_sku_upper_bound[self._product2storage[product_id]][product_id]
+        remaining_quota = max(upper_bound - self._product_level[product_id], 0)
+
+        return min(remaining_space, remaining_quota)
 
-            self._product_level[product_id] += unload_quantity
-            unloaded_quantities[product_id] = unload_quantity
+    def _add_product(self, product_id: int, quantity: int) -> None:
+        assert self._remaining_space_dict[self._product2storage[product_id]] >= quantity
+        self._product_level[product_id] += quantity
+        self._product_level_changed[product_id] = True
+        self._remaining_space_dict[self._product2storage[product_id]] -= quantity
 
-            self._changed_product_cache[product_id] = True
+    def _take_product(self, product_id: int, quantity: int) -> None:
+        assert self._product_level[product_id] >= quantity
+        self._product_level[product_id] -= quantity
```

Comment:
(Optional) Use `self._add_product(product_id, -quantity)` to replace line 156-158.

Reply:
if call `self._add_product` the assertion in line 149 would fail.

Since it is not a complex function, I prefer to use detailed and straight logic here (as now).